core-tests/shared/src/test/scala/zio/stm/TSemaphoreSpec.scala

```diff
@@ -0,0 +1,94 @@
+import zio.test._
+
+object TSemaphoreSpec extends ZIOBaseSpec {
+  override def spec = suite("TSemaphore")(
+    suite("factories")(
+      testM("make") {
+        checkM(Gen.long(1L, Int.MaxValue)) { expected =>
+          val actual = for {
+            sem <- TSemaphore.make(expected)
+            cap <- sem.available
+          } yield cap
+
+          assertM(actual.commit)(equalTo(expected))
+        }
+      }
+    ),
+    suite("acquire and release")(
+      testM("acquiring and releasing a permit should not change the availability") {
+        checkM(Gen.long(1L, Int.MaxValue)) { expected =>
+          val actual = for {
+            sem <- TSemaphore.make(expected)
+            _   <- sem.acquire *> sem.release
+            cap <- sem.available
+          } yield cap
+          assertM(actual.commit)(equalTo(expected))
+        }
+      },
+      testM("used capacity must be equal to the # of acquires minus # of releases") {
+        checkM(usedCapacityGen) {
+          case (capacity, acquire, release) =>
+            val actual = for {
+              sem <- TSemaphore.make(capacity)
+              _   <- repeat(sem.acquire)(acquire) *> repeat(sem.release)(release)
+              cap <- sem.available
+            } yield cap
+
+            val usedCapacity = acquire - release
+            assertM(actual.commit)(equalTo(capacity - usedCapacity))
+        }
+      },
+      testM("acquireN/releaseN(n) is acquire/release repeated N times") {
+        checkM(Gen.long(1, 100)) {
+          capacity =>
+            def acquireRelease(
+              sem: TSemaphore
+            )(acq: Long => STM[Nothing, Unit])(rel: Long => STM[Nothing, Unit]): STM[Nothing, (Long, Long)] =
+              for {
+                _            <- acq(capacity)
+                usedCapacity <- sem.available
+                _            <- rel(capacity)
+                freeCapacity <- sem.available
+              } yield (usedCapacity, freeCapacity)
+
+            STM.atomically {
+              for {
+                sem               <- TSemaphore.make(capacity)
+                acquireReleaseN   = acquireRelease(sem)(sem.acquireN)(sem.releaseN)
+                acquireReleaseRep = acquireRelease(sem)(repeat(sem.acquire))(repeat(sem.release))
+                resN              <- acquireReleaseN
+                resRep            <- acquireReleaseRep
+              } yield assert(resN)(equalTo(resRep)) && assert(resN)(equalTo((0L, capacity)))
+            }
+        }
+      },
+      testM("withPermit automatically acquires and releases a permit once an action completes") {
+        val expected = 1L
+        STM.atomically {
+          for {
+            sem    <- TSemaphore.make(expected)
+            _      <- sem.withPermit(STM.unit)
+            actual <- sem.available
+          } yield assert(actual)(equalTo(expected))
+        }
+      }
+    )
+  )
+
+  private def repeat[E, A](stm: STM[E, A])(n: Long): STM[E, A] = n match {
```

Comment:
Not as part of this PR but we should think about how to implement functionality like this in `STM` itself. In ZIO it is powered by `Schedule` which we have issues with using because it can execute arbitrary effects but maybe we could have a `TSchedule` structure more similar to the old encoding of `Schedule` that we could use for logic like this.